pkg/playback/http_playback.go

```diff
@@ -0,0 +1,413 @@
+	w.WriteHeader(statusCode)
+	fmt.Fprintf(w, "%v\n", err)
+	fmt.Printf("\n<-- %d error: %v\n", statusCode, err)
+}
+
+// These constants define the modes the playback server can be in
+const (
+	// in auto mode, the server records a new cassette if the given file initially doesn't exist
+	// and replays a cassette of the file does exist.
+	Auto   string = "auto"
+	Record string = "record"
+	Replay string = "replay"
+)
+
+// A Server implements the full functionality of `stripe playback` as a HTTP server.
+// Acting as a proxy server for the Stripe API, it can both record and replay interactions using cassette files.
+type Server struct {
+	httpRecorder recordServer
+	httpReplayer replayServer
+
+	remoteURL         string
+	cassetteDirectory string // absolute path to the root directory for all cassette filepaths
+
+	// state machine state
+	mode                  string // the user specified state (auto, record, replay)
+	isRecordingInAutoMode bool   // internal state used when in auto mode to keep track of the state for the current cassette (either recording or replaying)
+	cassetteLoaded        bool
+}
+
+// NewServer instantiates a Server struct, representing the configuration and current state of a playback proxy server
+// The cassetteDirectory param must be an absolute path
+// initialCasssetteFilepath can be a relative path (inretpreted relative to cassetteDirectory) or an absolute path
+func NewServer(remoteURL string, webhookURL string, absCassetteDirectory string, mode string, initialCassetteFilepath string) (server *Server, err error) {
+	server = &Server{}
+
+	// initialize server.httpRecorder and server.httpReplayer first, since calls to methods like
+	// server.loadCassette reference them.
+	server.httpRecorder = newRecordServer(remoteURL, webhookURL)
+	server.httpReplayer = newReplayServer(webhookURL)
+	server.remoteURL = remoteURL
+
+	err = server.switchMode(mode)
+	if err != nil {
+		return server, err
+	}
+
+	err = server.loadCassette(initialCassetteFilepath)
+	if err != nil {
+		return server, err
+	}
+
+	err = server.setCassetteDir(absCassetteDirectory)
+	if err != nil {
+		return server, err
+	}
+
+	return server, nil
+}
+
+// InitializeServer sets up and returns a http.Server that acts as a playback proxy
+func (rr *Server) InitializeServer(address string) *http.Server {
+	customMux := http.NewServeMux()
+	server := &http.Server{Addr: address, Handler: customMux}
+
+	// --- Webhook endpoint
+	customMux.HandleFunc("/playback/webhooks", rr.webhookHandler)
+
+	// --- Server control handlers
+	customMux.HandleFunc("/playback/mode/", func(w http.ResponseWriter, r *http.Request) {
+		// get mode
+		modeString := strings.TrimPrefix(r.URL.Path, "/playback/mode/")
+
+		wasCassetteLoaded := rr.cassetteLoaded
+		err := rr.switchMode(modeString)
+		if wasCassetteLoaded {
+			fmt.Println("/playback/mode: unloaded the cassette. Please load a new cassette before recording/replaying any new interactions.")
```

Comment:
(Non-blocking suggestion) WDYT about doing the logging consistent with how we do it in e.g. `stripe listen`? That is, we log with a log.Logger object, rather than writing explicitly to stdout with `fmt.Print*`. This would let us more easily introduce a "debug" log level if necessary, and I think it also makes it easier to add support for a `--json` flag

https://github.com/stripe/stripe-cli/blob/master/pkg/proxy/proxy.go#L68

Reply:
(This is a pretty big refactor, maybe best to do in a follow-up PR)

Reply:
Another related change made it make sense to this in this PR - switched over to logrus.Logger (which `stripe listen` uses) in feacb33396bd5fb09a3feef1749ea5d61eba03c3.